File: check_cases.clj
(defn euclid
  "Compute the greatest common divisor using Euclid's algorithms."
  [a b]
  (assert (and (integer? a) (integer? b)) "Expecting integers.")
  (loop [a a b b]
    (if (= b 0)
      (Math/abs a)
      (recur b (mod a b)))))

(defn gcd 
  "Compute the greatest common divisor of integers."
  [& rest]
  (reduce euclid 0 rest))

(defn units
  "Compute the units modulo m."
  [m]
  (assert (and (integer? m) (> m 0)) "Need positive integer.")
  (filter #(= (gcd % m) 1) (range m)))

(defn self-inverse
  "Check whether 'a' is its own inverse modulo 'm'."
  [a m]
  (assert (integer? a) "Need integer.")
  (= (mod (* a a) m) 1))

(defn all-inverses-self
  "Check whether all units modulo 'm' are their own inverses."
  [m]
  (assert (and (integer? m) (> m 0)) "Need positive integer.")
  (every? #(self-inverse % m) (units m)))

(def result
  "All numbers 'm' (except 1) where every unit modulo 'm' is its on inverse."
  (filter all-inverses-self (range 1 49)))

(println result)

File: check_cases.hs
-- The list of units modulo 'm'.
units :: Integer -> [Integer]
units m | m > 0 = [ x | x <- [1..m], gcd x m == 1 ]

-- Is 'a' its own inverse modulo 'm'?
autoInverse :: Integer -> Integer -> Bool
autoInverse m a | m > 0 = a^2 `mod` m == 1

-- Is every unit modulo 'm' its own inverse?
everyInverseAuto :: Integer -> Bool
everyInverseAuto m | m > 0 = all (autoInverse m) $ units m

-- All 'm' such that the units modulo 'm' are their own inverses.
result :: [Integer]
result = [ m | m <- [1..49], everyInverseAuto m ]

-- Print the result
main :: IO ()
main = putStrLn $ show result
